FlashIAP driver: Add retries to erase and program operations.

davidsaada opened the pull request:
### Description
Few boards may fail the write actions due to HW limitations (like critical drivers that
disable flash operations). Just retry a few times until success.
In addition, remove the redundant retries in NVStore (not needed now).

This fixes failures we saw in client tests, mainly with the UBLOX_EVK_ODIN_W2 board.
Tested with all flash related tests (FlashIAP, NVStore, KVStore) on K64F and UBLOX_EVK_ODIN_W2.

### Pull request type
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

ciarmcom commented:
@davidsaada, thank you for your changes.
@ARMmbed/mbed-os-storage @ARMmbed/mbed-os-core @ARMmbed/mbed-os-maintainers please review.

dannybenor commented on drivers/FlashIAP.cpp:
```diff
@@ -185,21 +195,27 @@ int FlashIAP::erase(uint32_t addr, uint32_t size)
-        if (ret != 0) {
-            ret = -1;
-            break;
+    // Few boards may fail the write actions due to HW limitations (like critical drivers that
```
Please fix the comment to "erase" instead of "write"

davidsaada replied:
fixed.

offirko commented on drivers/FlashIAP.cpp:
```diff
@@ -107,50 +109,58 @@ int FlashIAP::program(const void *buffer, uint32_t addr, uint32_t size)
+    // Few boards may fail the write actions due to HW limitations (like critical drivers that
+    // disable flash operations). Just retry a few times until success.
+    ret = -1;
+    for (unsigned int retry = 0; retry < num_write_retries && ret; retry++) {
```
should be an exit trigger of retries loop in case of success (maybe if total_size == 0)?

davidsaada replied:
ret is the exit trigger (ret == 0 will exit on the first iteration).

offirko replied:
sorry, I missed ret at the condition.
Could flash_program_page() be called several times for one "program" ?
If so, what happens if  not the first one fails , will it retry from the beginning?

davidsaada replied:
Very good point. It will retry from the beginning. This will likely succeed in most of the flash components (programming the same value twice), if not all, but there's always a chance that this logic won't always work. Will try to modify the code accordingly.

davidsaada replied:
Modified the code to retry the physical operation (and not on the entire process). Please re-review.

offirko replied:
Liked the idea (-:

Commit:
FlashIAP driver: Add retries to erase and program operations.

Few boards may fail the write actions due to HW limitations (like critical
drivers that disable flash operations). Just retry a few times until success.
In addition, remove the redundant retries in NVStore (not needed now).

offirko reviewed (approved):
LGTM

JanneKiiskila commented:
Client test results looking ok, please get this to 5.11.3 ASAP. @adbridge  @0xc0170

deepikabhavnani reviewed (approved):
Changes look fine to me.

For my understanding:
Issue was noticed earlier with NVStore and fix is now made generic for FlashIAP as I understand. But any reason for magic number 16 (I am happy that we are limiting to some number and not in while(1)) just curious.

Also description states failure is due to HW limitation like critical drivers that disable flash operations, do we have more information on this? Which driver disables flash operation in Mbed OS.

cmonr commented:
CI started

davidsaada commented:
> 
> 
> Changes look fine to me.
> 
> For my understanding:
> Issue was noticed earlier with NVStore and fix is now made generic for FlashIAP as I understand. But any reason for magic number 16 (I am happy that we are limiting to some number and not in while(1)) just curious.
> 
> Also description states failure is due to HW limitation like critical drivers that disable flash operations, do we have more information on this? Which driver disables flash operation in Mbed OS.

@deepikabhavnani We started noticing this issue in NVStore, but were afraid to make the change in the  driver back then, which was a mistake apparently. We saw it happening with WiFi (which is also what the client test application operated in the ODIN board), but then learned that this problem could also happen in other protocols such as BLE. 
As for the retries number - it just seemed like a reasonable one. In practice, we saw that 1-2 retires were enough, but couldn't tell whether other protocols would block the driver for longer times, so 16 seemed like a good number.

0xc0170 commented:
> Also description states failure is due to HW limitation like critical drivers that disable flash operations, do we have more information on this? Which driver disables flash operation in Mbed OS.

I would like to also understand this. So some stacks just disable these operations and led to failures in kvstore.

davidsaada commented:
> 
> 
> > Also description states failure is due to HW limitation like critical drivers that disable flash operations, do we have more information on this? Which driver disables flash operation in Mbed OS.
> 
> I would like to also understand this. So some stacks just disable these operations and led to failures in kvstore.

Was KVStore in this case (discovered by client tests), but the failures were lower - in the flash driver, hence the change.

cmonr commented:
https://github.com/ARMmbed/mbed-os/pull/9508#issuecomment-459155182
